### PR TITLE
Mobility Refactor

### LINKED ISSUE #8108: when drugged and you go unconcious, game thinks you're crit and when you try to whisper, you immediately die, even though the game says you are 100% health

### PR DESCRIPTION
**Pull request by Tsar-Salat**
Ports /tg/:
- ~~https://github.com/tgstation/tgstation/pull/49237~~
- ~~https://github.com/tgstation/tgstation/pull/49751~~
- ~~https://github.com/tgstation/tgstation/pull/50018~~
- ~~https://github.com/tgstation/tgstation/pull/50270~~
- ~~https://github.com/tgstation/tgstation/pull/50327~~
- ~~https://github.com/tgstation/tgstation/pull/51348~~
- ~~https://github.com/tgstation/tgstation/pull/52578~~
- ~~https://github.com/tgstation/tgstation/pull/52884~~
- https://github.com/tgstation/tgstation/pull/52929
- ~~https://github.com/tgstation/tgstation/pull/53117~~
- ~~https://github.com/tgstation/tgstation/pull/52892 (AFTER crit status refactor)~~
- ~~https://github.com/tgstation/tgstation/pull/53349~~
- https://github.com/tgstation/tgstation/pull/53374
- ~~https://github.com/tgstation/tgstation/pull/53981~~
- https://github.com/tgstation/tgstation/pull/54078
- https://github.com/tgstation/tgstation/pull/54183
- https://github.com/tgstation/tgstation/pull/54320
- https://github.com/tgstation/tgstation/pull/54411
- https://github.com/tgstation/tgstation/pull/55579

## About The Pull Request

Closes #8108 

Contributes to #7213 

This is an atomization and successor of #8474. 
I decided to make a new pr because I wasnt proud of the previous pr. At the time I was much less experienced, and mostly relied on the Monke port, rather than understanding the source material.

Lets get into it. The main concept behind TG's mobility refactor is to make mobility more event-based, reduce stupid is_var checks, and move all these proc-based mobility shit to signals/status effects. 
Doing so lets us append a LOT more behavior, it should lead to an easier time if we get to porting wounds or kapulimbs.

tl;dr the purpose of this is to kill the `/update_mobility` method, and compensate for the handling of that code instead of calling this one huge terrible method

## Why It's Good For The Game

image

This is an adequate summary of what this does for us.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Rest control
restcontrol

Stasis applies properly to plasmameme
stasis-plasmameme

I busted my arm
limb-disabling

cryo applies correctly
cryolooksfine

tucked in
tuckedin

</details>

## Changelog
:cl: RKz,  Rohesie, ExcessiveUseOfCobblestone
add: transitioning between resting and unresting is much more smooth, with an interruptable do_after
add: HUD feature that will allow you to succumb in Hardcrit. You dont have to type the ghost verb or whisper to kill yourself, just hit the HUD popup
add: with the addition of the succumb feature, you can no longer kill yourself accidentely while trying to whisper in a state of softcrit or hardcrit. You will speak, or not be able to in that instance.
tweak: Hardcrit is an actual state now, I moved most stuff that were intended for it from the unconscious state. The majority of checks will now check for both states however.
refactor: basically all the limb disabling system, makes it event-based
refactor: made a bunch of movement-related code signal-based instead of /proc/
code: carbon body position checks
code: made a bunch of overused redefines into setters
rscadd: ghost sprite for succumbing
/:cl:

### CODE REVIEW

**Comment by Tsar-Salat** [comment]
Oh yeah, this one also doesn't conflict with #8204, as I included no blindness changes

**Comment by Tsar-Salat** [comment]
mfw

File: code/_onclick/click.dm [hunk 1]
```diff
@@ -113,9 +113,9 @@
 		var/obj/mecha/M = loc
 		return M.click_action(A,src,params)
 
-	if(restrained())
+	if(HAS_TRAIT(src, TRAIT_HANDS_BLOCKED))
 		changeNext_move(CLICK_CD_HANDCUFFED)   //Doing shit in cuffs shall be vey slow
-		RestrainedClickOn(A)
+		UnarmedAttack(A)
 		return
```
Comment (PowerfulBacon) [inline code/_onclick/click.dm:119]: Will this not allow actions while cuffed, as well as allowing monkeys to attack_paw anything while cuffed?

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
This PR is going to be impossible to review, theres a lot of changes in here not related to mobility at all, such as succumb alerts.

Half the CL isn't even a refactoring of mobility:
add: HUD feature that will allow you to succumb in Hardcrit. You dont have to type the ghost verb or whisper to kill yourself, just hit the HUD popup
tweak: Hardcrit is an actual state now, I moved most stuff that were intended for it from the unconscious state. The majority of checks will now check for both states however.
refactor: basically all the limb disabling system, makes it event-based
code: made a bunch of overused redefines into setters

File: code/_onclick/click.dm (hunk 1 above)
Comment (Tsar-Salat) [reply code/_onclick/click.dm:119]: How? 

The trait blocks the use of your hands. It does functionally the same thing as before.

https://discordapp.com/channels/427337870844362753/943311132007997490/1107705086169321502

**Comment by Tsar-Salat** [comment]
Ah, I realize I forgot to add a screen alert icon, which made the alert not appear. 

Should be fixed now.

The target is in crit. A screen alert appears indicating they can succumb if they choose.
finsero

Clicking the alert allows them to say final words, or back out.
lastio

Commiting to succumbing
deathio

File: code/_onclick/click.dm (hunk 1 above)
Comment (Tsar-Salat) [reply code/_onclick/click.dm:119]: Also, All UnarmedAttack check for TRAIT_HANDS_BLOCKED, so there shouldnt be any issues.

**Comment by Tsar-Salat** [comment]
> tweak: Hardcrit is an actual state now, I moved most stuff that were intended for it from the unconscious state. The majority of checks will now check for both states however.
> refactor: basically all the limb disabling system, makes it event-based
> code: made a bunch of overused redefines into setters

Sure, I can move the succumbing to another PR, but all the rest are connected to update_mobility.
 
The whole aim of this PR was to kill update_mobility and decouple the multiple systems that were connected to it, that definitely applies to limb disabling, and checking health states.

Making redefines into setters allows you to append code to them.

**Comment by Tsar-Salat** [comment]
Enforced purity and linters in 7b033c9

**Comment by Tsar-Salat** [comment]
Merge basepixels before this. I'd like to purify some calls in this pr.

File: code/datums/components/footstep.dm [hunk 2]
```diff
@@ -98,7 +97,7 @@
 			//Sound of wearing shoes always plays, special movement sound
 			// IE (server motors wont play bare footed.)
 			if(H.dna.species.special_step_sounds)
-				playsound(T, pick(H.dna.species.special_step_sounds), 50, TRUE)
+				playsound(T, pick(H.dna.species.special_step_sounds), 30, TRUE)
```
Comment (itsmeow) [inline code/datums/components/footstep.dm]: wat. edit the volume of the ogg if its too loud but do it in another pr

File: code/datums/components/riding.dm [hunk 3]
```diff
@@ -71,7 +71,7 @@
 /datum/component/riding/proc/ride_check(mob/living/M)
 	var/atom/movable/AM = parent
 	var/mob/AMM = AM
-	if((ride_check_rider_restrained && M.restrained(TRUE)) || (ride_check_rider_incapacitated && M.incapacitated(FALSE, TRUE)) || (ride_check_ridden_incapacitated && istype(AMM) && AMM.incapacitated(FALSE, TRUE)))
```
Comment (itsmeow) [inline code/datums/components/riding.dm:74]: the ridden check is gone

File: code/datums/status_effects/debuffs.dm [hunk 4]
```diff
@@ -6,41 +8,99 @@
 	alert_type = null
 	var/needs_update_stat = FALSE
 
-/datum/status_effect/incapacitating/on_creation(mob/living/new_owner, set_duration, updating_canmove)
-	if(isnum_safe(set_duration))
```
Comment (itsmeow) [inline code/datums/status_effects/debuffs.dm:12]: seems bad

File: code/game/objects/effects/step_triggers.dm [hunk 5]
```diff
@@ -68,10 +68,8 @@
 		if(AM in T.affecting)
 			return
 
-	if(isliving(AM))
-		var/mob/living/M = AM
-		if(immobilize)
-			M.mobility_flags &= ~MOBILITY_MOVE
+	if(immobilize)
+		ADD_TRAIT(AM, TRAIT_IMMOBILIZED, src)
```
Comment (itsmeow) [inline code/game/objects/effects/step_triggers.dm]: this uses src but below uses REF(src)

**Comment by itsmeow** [comment]
What is the difference between TRAIT_CRITICAL_CONDITION and HARDCRIT stat?

File: code/modules/instruments/items.dm [hunk 6]
```diff
@@ -41,7 +41,7 @@
 	ui_interact(user)
 
 /obj/item/instrument/ui_interact(mob/living/user)
-	if(!isliving(user) || user.stat || (user.restrained() && !ispAI(user)))
```
Comment (itsmeow) [inline code/modules/instruments/items.dm:44]: this might break pAI shit

File: code/modules/mob/living/carbon/alien/larva/larva.dm [hunk 7]
```diff
@@ -65,3 +73,6 @@
 /mob/living/carbon/alien/larva/stripPanelEquip(obj/item/what, mob/who)
 	to_chat(src, "<span class='warning'>You don't have the dexterity to do this!</span>")
 	return
+
+/mob/living/carbon/alien/larva/canBeHandcuffed()
+	return TRUE
```
Comment (itsmeow) [inline code/modules/mob/living/carbon/alien/larva/larva.dm:78]: how the fuck do you handcuff a fucking worm

File: code/modules/mob/living/damage_procs.dm [hunk 8]
```diff
@@ -159,19 +159,21 @@
...
 /mob/living/proc/setOxyLoss(amount, updating_health = TRUE, forced = FALSE)
-	if(status_flags & GODMODE)
-		return 0
+	if(!forced && status_flags & GODMODE)
+		return
+	. = oxyloss
 	oxyloss = amount
 	if(updating_health)
 		updatehealth()
-	return amount
```
Comment (itsmeow) [inline code/modules/mob/living/damage_procs.dm:174]: does anything rely on this behavior of returning the amount passed in? because now it's returning the old oxyloss value instead

File: code/modules/mob/living/init_signals.dm [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_UI_BLOCKED), PROC_REF(on_ui_blocked_trait_loss))
+
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_PULL_BLOCKED), PROC_REF(on_pull_blocked_trait_gain))
+	RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_PULL_BLOCKED), PROC_REF(on_pull_blocked_trait_loss))
+
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_INCAPACITATED), PROC_REF(on_incapacitated_trait_gain))
+	RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_INCAPACITATED), PROC_REF(on_incapacitated_trait_loss))
+
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_RESTRAINED), PROC_REF(on_restrained_trait_gain))
+	RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_RESTRAINED), PROC_REF(on_restrained_trait_loss))
```
Comment (itsmeow) [inline code/modules/mob/living/init_signals.dm:31]: registering this many signals on yourself seems a little bit wasteful?

File: code/modules/mob/living/living.dm [hunk 10]
```diff
@@ -793,7 +847,7 @@
 
 	SEND_SIGNAL(src, COMSIG_LIVING_RESIST, src)
 	//resisting grabs (as if it helps anyone...)
-	if(!restrained(ignore_grab = 1) && pulledby)
+	if(!HAS_TRAIT(src, TRAIT_RESTRAINED) && pulledby)
```
Comment (itsmeow) [inline code/modules/mob/living/living.dm:848]: does this work?

File: code/modules/mob/living/living.dm [hunk 11]
```diff
@@ -878,7 +932,7 @@
 		animate(pixel_y = base_pixel_y - 2, time = 10, loop = -1, flags = ANIMATION_RELATIVE)
 		setMovetype(movement_type | FLOATING)
 	else if(((!on || fixed) && (movement_type & FLOATING)))
-		animate(src, pixel_y = base_pixel_y + get_standard_pixel_y_offset(lying), time = 1 SECONDS)
+		animate(src, pixel_y = get_standard_pixel_y_offset(lying_angle), time = 10)
```
Comment (itsmeow) [inline code/modules/mob/living/living.dm]: why this

**Comment by Tsar-Salat** [comment]
> What is the difference between TRAIT_CRITICAL_CONDITION and HARDCRIT stat?

They're just replacements for current code, to reduce snowflake checks.

In this case, the `HARD_CRIT` value in stat replaces the `InFullCritical()` proc, while `TRAIT_CRITICAL_CONDITION` replaces the usage of `InCritical()`.

The point of this replacement is to clean up code for the succumbing HUD icon.

tl;dr its just converting those two snowflake procs into traits that are checked, for the purpose of killing update_mobility and trying to convert code into traits and status effects.

It was added in this tg pr:
- https://github.com/tgstation/tgstation/pull/53117

File: code/modules/mob/living/say.dm [hunk 12]
```diff
@@ -99,16 +97,24 @@ GLOBAL_LIST_INIT(department_radio_keys, list(
...
 	var/datum/saymode/saymode = SSradio.saymodes[message_mods[RADIO_KEY]]
-	var/in_critical = InCritical()
 
 	if(!message)
 		return
 
 	message = check_for_custom_say_emote(message, message_mods)
 
-	if(stat == DEAD)
-		say_dead(original_message)
```
Comment (itsmeow) [inline code/modules/mob/living/say.dm:110]: this breaks the ability to speak in dchat if you're inside your dead body

File: code/modules/mob/living/status_procs.dm [hunk 13]
```diff
@@ -215,33 +215,28 @@
 		if(P)
 			P.duration += amount
 		else if(amount > 0)
-			P = apply_status_effect(STATUS_EFFECT_PARALYZED, amount, updating)
+			P = apply_status_effect(STATUS_EFFECT_PARALYZED, amount,   )
 		return P
 
 //Blanket
-/mob/living/proc/AllImmobility(amount, updating)
+/mob/living/proc/AllImmobility(amount,   )
```
Comment (itsmeow) [inline code/modules/mob/living/status_procs.dm]: this file is just full of empty arguments `amount,   )` don't do that

File: code/modules/mob/mob.dm [hunk 14]
```diff
@@ -494,23 +490,15 @@
 
 /mob/living/blind_examine_check(atom/examined_thing)
 	//need to be next to something and awake
-	if(!Adjacent(examined_thing) || incapacitated())
+	if(!in_range(examined_thing, src) || incapacitated())
```
Comment (itsmeow) [inline code/modules/mob/mob.dm]: is this correct?

File: code/modules/projectiles/projectile.dm [hunk 15]
```diff
@@ -493,9 +493,9 @@
 			return FALSE
 		if(!L.density)
 			return FALSE
-		if(!L.lying)
+		if(L.body_position != LYING_DOWN)
 			return TRUE
-		var/stunned = !CHECK_BITFIELD(L.mobility_flags, MOBILITY_USE | MOBILITY_STAND | MOBILITY_MOVE)
+		var/stunned = HAS_TRAIT(L, TRAIT_IMMOBILIZED) && HAS_TRAIT(L, TRAIT_FLOORED) && HAS_TRAIT(L, TRAIT_HANDS_BLOCKED)
```
Comment (itsmeow) [inline code/modules/projectiles/projectile.dm:501]: shouldn't this use OR? or like be inverted?

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 16]
```diff
@@ -227,19 +229,48 @@
...
 		return
 
-	brute_dam	= round(max(brute_dam - brute, 0), DAMAGE_PRECISION)
-	burn_dam	= round(max(burn_dam - burn, 0), DAMAGE_PRECISION)
-	stamina_dam = round(max(stamina_dam - stamina, 0), DAMAGE_PRECISION)
-	if(owner && updating_health)
-		owner.updatehealth()
-	if(owner.dna && owner.dna.species && (REVIVESBYHEALING in owner.dna.species.species_traits))
-		if(owner.health > 0 && !owner.ishellbound())
-			owner.revive(0)
```
Comment (itsmeow) [inline code/modules/surgery/bodyparts/bodyparts.dm:237]: this feature appears removed

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 17]
```diff
@@ -249,28 +280,123 @@
...
+	if(!can_be_disabled)
+		set_disabled(FALSE)
+		CRASH("update_disabled called with can_be_disabled false")
 
-/obj/item/bodypart/proc/is_disabled()
 	if(HAS_TRAIT(src, TRAIT_PARALYSIS))
-		return BODYPART_DISABLED_PARALYSIS
-	if(can_dismember() && !HAS_TRAIT(owner, TRAIT_NOLIMBDISABLE))
-		. = disabled //inertia, to avoid limbs healing 0.1 damage and being re-enabled
-		if((get_damage(TRUE) >= max_damage) || (HAS_TRAIT(owner, TRAIT_EASYLIMBDISABLE) && (get_damage(TRUE) >= (max_damage * 0.6)))) //Easy limb disable disables the limb at 40% health instead of 0%
```
Comment (itsmeow) [inline code/modules/surgery/bodyparts/bodyparts.dm:259]: is this trait not used anymore?

File: code/modules/surgery/bodyparts/bodyparts.dm [hunk 18]
```diff
@@ -610,32 +768,62 @@
...
-		return BODYPART_DISABLED_PARALYSIS
-	return ..()
+/obj/item/bodypart/r_arm/set_owner(new_owner)
+	. = ..()
+	if(. == FALSE)
+		return
+	if(owner)
+		if(HAS_TRAIT(owner, TRAIT_PARALYSIS_R_ARM))
+			ADD_TRAIT(src, TRAIT_PARALYSIS, TRAIT_PARALYSIS_R_ARM)
+			RegisterSignal(owner, SIGNAL_REMOVETRAIT(TRAIT_PARALYSIS_R_ARM), PROC_REF(on_owner_paralysis_loss))
```
Comment (itsmeow) [inline code/modules/surgery/bodyparts/bodyparts.dm:785]: this behavior seems kinda copypasta, can you make it into a var on the bodypart type for like "unique paralysis trait"

File: code/_onclick/click.dm (hunk 1 above)
Comment (Tsar-Salat) [reply code/_onclick/click.dm:119]: marking as addressed

File: code/modules/surgery/bodyparts/head.dm [hunk 19]
```diff
@@ -93,7 +93,7 @@
 
 
 /obj/item/bodypart/head/can_dismember(obj/item/I)
-	if(!((owner.stat == DEAD) || owner.InFullCritical()))
+	if(owner && owner.stat <= HARD_CRIT)
```
Comment (itsmeow) [inline code/modules/surgery/bodyparts/head.dm]: should this check the trait instead?

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by itsmeow** [comment]
I *really* wish this was more atomized as it's fuckin hell to review and is just gonna have to get the old YOLO-TM

**Comment by Tsar-Salat** [comment]
> I _really_ wish this was more atomized as it's fuckin hell to review

Understood. 

Im keeping this in draft for now as a sort of end-product of what I want mobility to look like. Ill eventually turn this PR into just the deathknell for update_mobility, and attempt to offload stuff that isnt directly involved in doing so into some atomizations, like #9291

File: code/datums/components/footstep.dm (hunk 2 above)
Comment (Tsar-Salat) [reply code/datums/components/footstep.dm]: reverted in 564325b

File: code/game/objects/effects/step_triggers.dm (hunk 5 above)
Comment (Tsar-Salat) [reply code/game/objects/effects/step_triggers.dm]: fixed in aa31e5e

File: code/modules/mob/living/status_procs.dm (hunk 13 above)
Comment (Tsar-Salat) [reply code/modules/mob/living/status_procs.dm]: cleaned up in 88c0d49

File: code/modules/mob/mob.dm (hunk 14 above)
Comment (Tsar-Salat) [reply code/modules/mob/mob.dm]: Nope. fixed in 7faf35e

File: code/modules/mob/living/say.dm (hunk 12 above)
Comment (Tsar-Salat) [reply code/modules/mob/living/say.dm:110]: Nope.
image

File: code/modules/surgery/bodyparts/head.dm (hunk 19 above)
Comment (Tsar-Salat) [reply code/modules/surgery/bodyparts/head.dm]: No need. Hardcrit stat replaced it here.

File: code/modules/surgery/bodyparts/bodyparts.dm (hunk 16 above)
Comment (Tsar-Salat) [reply code/modules/surgery/bodyparts/bodyparts.dm:237]: Good catch. readded 6d4d32b

File: code/modules/surgery/bodyparts/bodyparts.dm (hunk 17 above)
Comment (Tsar-Salat) [reply code/modules/surgery/bodyparts/bodyparts.dm:259]: see /obj/item/bodypart/proc/set_owner

File: code/modules/mob/living/carbon/alien/larva/larva.dm (hunk 7 above)
Comment (Tsar-Salat) [reply code/modules/mob/living/carbon/alien/larva/larva.dm:78]: > how the fuck do you handcuff a fucking worm

its not new behavior. 
image

You can already do it in the code rn 🤷

File: code/modules/mob/living/init_signals.dm (hunk 9 above)
Comment (Tsar-Salat) [reply code/modules/mob/living/init_signals.dm:31]: I thought it was used to improve performance in making things event-based? https://github.com/tgstation/tgstation/pull/74084

**Comment by Tsar-Salat** [comment]
Two down! Nice.

**Comment by Tsar-Salat** [comment]
im taking a break, Semester begins soon and this is affecting my mental health
